[`pylint`] Update `missing-maxsplit-arg` docs and error to suggest proper usage (`PLC0207`)

junhsonjb opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary

<!-- What's the purpose of the change? What does it do, and why? -->

Fix #18383 by updating the documentation and error message to explain that users should use `rsplit` in order to access the last element of the result with `maxsplit=1`

## Test Plan

<!-- How was it tested? -->

Only documentation and an error message was changed. As such, snapshots were updated to reflect the new error message. With this change, all existing tests pass.

junhsonjb commented:
Thanks for the comments, that makes sense! I just patched some updates to address them!

MeGaGiGaGon commented on crates/ruff_linter/src/rules/pylint/rules/missing_maxsplit_arg.rs:
```diff
@@ -29,14 +29,37 @@ use crate::checkers::ast::Checker;
+        let correct_function = match index {
+            0 => "split",
+            -1 => "rsplit",
+            _ => "", // We should never hit this case
```
This should probably be an `unreachable!` instead, since if that case is hit it should cause a failure instead of a silent issue.

ntBre replied:
I think we should probably define a simple enum instead of using an `i64`. That way we can check for known values up front and bail out of the whole diagnostic. Then we can `match` exhaustively here.

MeGaGiGaGon commented on crates/ruff_linter/src/rules/pylint/snapshots/ruff_linter__rules__pylint__tests__PLC0207_missing_maxsplit_arg.py.snap:
```diff
@@ -21,7 +21,7 @@ missing_maxsplit_arg.py:15:1: PLC0207 Accessing only the first or last element o
    |
 
-missing_maxsplit_arg.py:16:1: PLC0207 Accessing only the first or last element of `str.split()` without setting `maxsplit=1`
+missing_maxsplit_arg.py:16:1: PLC0207 Instead of `str.rsplit()`, call `str.split()` and pass `maxsplit=1`
```
I think you need to have both the split type and split index in `MissingMaxsplitArg`, since this is currently making the message backwards (first should stay as `split`, second should recommend switching to `rsplit`)

junhsonjb replied:
Just so that I'm clear, which cases do you mean by "first" and "second"? This particular error message (line 24 of the `.snap` file) recommends `split` because the python code is accessing the first element. With `maxsplit=1`, we need to call `split` or else we'd get the wrong output.

`rsplit` would give us something like this:
```python
```

MeGaGiGaGon replied:
Condensed down to just the error and message:

```snap
14 | "1,2,3".split(",")[0]  # [missing-maxsplit-arg]
   | ^^^^^^^^^^^^^^^^^^^^^ PLC0207
missing_maxsplit_arg.py:15:1: PLC0207 Instead of `str.split()`, call `str.rsplit()` and pass `maxsplit=1`


15 | "1,2,3".split(",")[-1]  # [missing-maxsplit-arg]
   | ^^^^^^^^^^^^^^^^^^^^^^ PLC0207
missing_maxsplit_arg.py:16:1: PLC0207 Instead of `str.rsplit()`, call `str.split()` and pass `maxsplit=1`
```
The messages are backwards. `"1,2,3".split(",")[0]` should stay as `split`, but the message says to use `rsplit`. `"1,2,3".split(",")[-1]` should switch to `rsplit`, but the message says to use `split` instead of `rsplit` dispite the code not having `rsplit` to begin with.

MeGaGiGaGon replied:
Yep, you're right, I was misreading it

junhsonjb commented:
Hey @ntBre — just wanted to check in on this. I’ve pushed some updates addressing your feedback above. When you get the chance, let me know if there’s anything else you’d like me to adjust. Thanks again! 🙂

ntBre reviewed (approved):
This looks great, thank you! And thanks for the ping.

I'm pretty sure these changes give us all the information we need to generate an autofix too. Feel free to work on that in a follow-up PR if you're interested :)

junhsonjb commented:
> This looks great, thank you! And thanks for the ping.
> 
> I'm pretty sure these changes give us all the information we need to generate an autofix too. Feel free to work on that in a follow-up PR if you're interested :)

@ntBre Great idea! PR is [here](https://github.com/astral-sh/ruff/pull/19387)
